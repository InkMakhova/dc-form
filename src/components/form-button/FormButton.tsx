import { JSX } from 'react'
import { Button } from 'react-aria-components'
import styles from './FormButton.module.css'
import { FormButtonProps } from './FormButton.types.ts'

function FormButton(props: FormButtonProps): JSX.Element {
  const {
    type,
    size,
    icon,
    iconPosition,
    text,
    ariaLabel,
    onClick,
    disabled,
    slot
  } = props;

  const isIconOnly: boolean = !!icon && !text;

  // Render
  return (
    <Button
      className={`
        ${styles.button}
        ${styles[`button--${size}`]}
        ${styles[`button--${type}`]}
        ${isIconOnly ? styles['button--icon-only'] : ''}
        ${slot ? (slot === 'decrement' ? styles['button--decrement'] : styles['button--increment']) : ""}
      `}
      isDisabled={disabled}
      aria-label={ariaLabel}
      slot={slot}
      onClick={onClick}
    >
      { icon &&
        <span
          className={`
            ${styles.button__icon} 
            ${iconPosition ? styles[`button__icon--${iconPosition}`] : ''}
          `}
        >
          {icon}
        </span>
      }
      { text }
    </Button>
  )
}

export default FormButton
