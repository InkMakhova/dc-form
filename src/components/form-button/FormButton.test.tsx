import { fireEvent, render, screen } from '@testing-library/react'
import {  expect, vitest} from 'vitest'
import '@testing-library/jest-dom'
import DeleteIcon from '../../assets/delete.svg?react'
import { ButtonSize, ButtonVariant } from './FormButton.types.ts'
import FormButton from './FormButton'

test('Button renders with text', () => {
  render(
    <FormButton
      type="button"
      text="Submit"
      size={ButtonSize.Medium}
      variant={ButtonVariant.PrimaryContained}
      ariaLabel="Submit"
    />
  )

  expect(screen.getByText("Submit")).toBeInTheDocument();
});

test('Button renders with icon only', () => {
  render(
      <FormButton
        type="button"
        icon={<DeleteIcon />}
        ariaLabel="Delete"
        size={ButtonSize.Small}
        variant={ButtonVariant.PrimaryOutlined}
      />
  )
  expect(screen.getByLabelText("Delete")).toBeInTheDocument();
});

test('Button applies correct classes for size and variant', () => {
  render(
    <FormButton
      type="button"
      text="Test"
      size={ButtonSize.Medium}
      variant={ButtonVariant.PrimaryContained}
      ariaLabel="Test"
    />);

  const button = screen.getByTestId('form-button');
  expect(button.className).toMatch(/button--medium/);
  expect(button.className).toMatch(/button--primary-contained/);
});

test('Button applies icon-only class if no text is provided', () => {
  render(
    <FormButton
      type="button"
      icon={<DeleteIcon />}
      ariaLabel="Delete"
      size={ButtonSize.Small}
      variant={ButtonVariant.PrimaryOutlined}
    />
  );
  const button = screen.getByTestId('form-button');
  expect(button.className).toMatch(/button--icon-only/);
});

test('Button calls onPress when clicked', () => {
  const handlePress = vitest.fn();
  render(
    <FormButton
      type="button"
      text="Click"
      onPress={handlePress}
      size={ButtonSize.Medium}
      variant={ButtonVariant.PrimaryContained}
      ariaLabel="Click"
    />
  );
  fireEvent.click(screen.getByTestId('form-button'));
  expect(handlePress).toHaveBeenCalledTimes(1);
});

test('Button does not call onPress when disabled', () => {
  const handlePress = vitest.fn();
  render(
    <FormButton
      type="button"
      text="Click"
      onPress={handlePress}
      disabled
      size={ButtonSize.Medium}
      variant={ButtonVariant.PrimaryContained}
      ariaLabel="Click"
    />);
  fireEvent.click(screen.getByTestId('form-button'));
  expect(handlePress).not.toHaveBeenCalled();
});
