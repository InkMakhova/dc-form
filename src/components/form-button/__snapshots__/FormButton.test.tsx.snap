// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div>
  <button
    aria-label="Click"
    class="
        _button_4e4ded
        _button--medium_4e4ded
        _button--primary-contained_4e4ded
        
      "
    data-disabled="true"
    data-rac=""
    data-testid="form-button"
    disabled=""
    id="react-aria-«rc»"
    type="button"
  >
    Click
  </button>
</div>
`;
